quickwit/quickwit-search/src/collector.rs

```diff
@@ -635,9 +794,65 @@ impl SegmentPartialHit {
     }
 }
 
+const BUFFER_LEN: usize = tantivy::COLLECT_BLOCK_BUFFER_LEN;
+/// Store the filtered docs in `filtered_docs_buffer` if `timestamp_filter_opt` is present.
+///
+/// Returns the number of docs.
+///
+/// Ideally we would return just final docs slice, but we can't do that because of the borrow
+/// checker.
+fn compute_filtered_block<'a>(
+    timestamp_filter_opt: &Option<TimestampFilter>,
+    docs: &'a [DocId],
+    filtered_docs_buffer: &'a mut [DocId; BUFFER_LEN],
+) -> usize {
+    let Some(timestamp_filter) = &timestamp_filter_opt else {
+        return docs.len();
+    };
+    let mut len = 0;
+    for &doc in docs {
+        filtered_docs_buffer[len] = doc;
+        len += if timestamp_filter.is_within_range(doc) {
+            1
+        } else {
+            0
+        };
+    }
+    len
+}
+
 impl SegmentCollector for QuickwitSegmentCollector {
     type Fruit = tantivy::Result<LeafSearchResponse>;
 
+    #[inline]
+    fn collect_block(&mut self, docs: &[DocId]) {
+        let num_docs =
+            compute_filtered_block(&self.timestamp_filter_opt, docs, &mut self.filtered_docs);
+
+        // Update results
+        self.num_hits += num_docs as u64;
+
+        if self.top_k_hits.max_len() != 0 {
+            self.collect_top_k_block(num_docs, docs);
```

Comment:
At this point, docs has not gone through the timestamp filter. Did you mean to put it one line below?